src/main/java/com/hubspot/jinjava/interpret/Context.java

```diff
@@ -290,6 +291,21 @@ public void handleDeferredNode(Node node) {
     return ImmutableSet.copyOf(deferredNodes);
   }
 
+  public void handlePreservedRawTag() {
+    hasPreservedRawTags = true;
+    if (getParent() != null) {
+      Context parent = getParent();
+      //Ignore global context
+      if (parent.getParent() != null) {
+        getParent().handlePreservedRawTag();
+      }
+    }
+  }
+
+  public boolean getHasPreservedRawTags() {
```

Comment:
I don't see this function is used. Is this needed?

Reply:
It's for rendering implementations such as with HubSpot's internal pre-renderer

Reply:
I think I am missing the whole picture. Is this the case,  the caller set the config to preserve raw tags, later it wants to confirm it did render with the flag set?

Reply:
So if the caller sets to preserve raw tags, but there are no raw tags to preserve and no deferred values, the pre-rendered template can be served as-is.

If however, there are any deferred values, then the second render will also get deferred until later.

Finally, if there are no deferred values, but raw tags are preserved, then the second-pass should happen immediately with the flag off to remove the raw tags and the result can be served as-is.

This is because there isn't a way of knowing if there are going to be deferred values down the line so it must always preserve raw tags if the flag is set. 3 examples assuming `deferred` is deferred:

Example 1, no deferred values, but a raw tag is preserved:
---
```
{% raw %}{{ foo.bar }}{% endraw %}
{{ resolved.variable }}
```
With `preserveRawTags=true` the output of the first render is:
```
{% raw %}{{ foo.bar }}{% endraw %}
I am resolved.
```
Since there are no deferred values, it should be rendered again with `preserveRawTags=false`. The output of the second render is:
```
{{ foo.bar }}
I am resolved.
```

Example 2, there are no deferred values or raw tags preserved:
---
```
Hello
{{ resolved.variable }}
```
With `preserveRawTags=true` the output of the first render is:
```
Hello
I am resolved.
```
Since there are no deferred values or preserved tags, the output of the first render can be served as-is.

Example 3, there is a deferred value and preserved raw tag:
---
```
{% raw %}{{ foo.bar }}{% endraw %}
{{ deferred.variable }}
```
With `preserveRawTags=true` the output of the first render is:
```
{% raw %}{{ foo.bar }}{% endraw %}
{{ deferred.variable }}
```
Since there is a deferred variable, it should be rendered again at serve-time with `preserveRawTags=false`. The output of the second render is:
```
{{ foo.bar }}
I am resolved.
```

Reply:
Thanks for the detailed examples. It helps a lot. But I still don't see where it needs to call `getHasPreservedRawTags()`. If the caller set it right before then calling the rendering, it should know it (local variable, etc).

Reply:
I think you're mistaking the config option for this. The config option tells the interpreter to preserve the raw tags. `getHasPreservedRawTags()` is only `true` if a raw tag was encountered and preserved. If there were no raw tags, then `getHasPreservedRawTags()` will be `false`.

Essentially, the method lets the caller know if any raw tags were preserved. It prevents an unnecessary second pass from happening. If this method didn't exist, it wouldn't break anything, but **Example 2** would have an extra rendering call that would yield the same result, which can be avoided by using this method.

There isn't a method to set `private boolean hasPreservedRawTags`, it is only set in the context when `handlePreservedRawTag()` gets called.

Reply:
Oh, my bad. I mis-read it. I am OK with this. However, if we do not have this, as you said, in example 2, we would have an extra rendering call -- it should be fast because there is nothing to render.

Reply:
That is true. I think it would be a minimal tradeoff of speed for more simplicity, so I'll just take this out.
This has shown me that it is too confusing to be worth the minimal performance gain